src/libertem/common/shape.py

```diff
@@ -151,6 +151,42 @@ def __eq__(self, other):
         values_eq = tuple(self) == tuple(other)
         return dims_eq and values_eq
 
+    def __add__(self, other):
+        """
+        Addition is supported between a Shape object and a tuple.
+        Either the navigation or the signal dimensions of the Shape object must be zero
+        and the tuple is added to the non-zero dimension.
+        """
+        if isinstance(other, tuple):
+            if self._sig_dims != 0 and len(self._nav_shape) != 0:
+                raise ValueError("To add a Shape object to a tuple, either the navigation "
+                                 "or the signal dimensions of the Shape object must be zero")
+            elif self._sig_dims == 0 and len(self._nav_shape) == 0:
+                raise ValueError("Shape object is empty")
+            elif self._sig_dims == 0:
+                return Shape(self._nav_shape + other, sig_dims=0)
+            else:
+                return Shape(self._sig_shape + other, sig_dims=len(self._sig_shape + other))
+        else:
+            return NotImplemented
+
+    def __radd__(self, other):
+        """
+        Reflected addition of a Shape object and a tuple
+        """
+        if isinstance(other, tuple):
+            if self._sig_dims != 0 and len(self._nav_shape) != 0:
+                raise ValueError("To add a Shape object to a tuple, either the navigation "
+                                 "or the signal dimensions of the Shape object must be zero")
+            elif self._sig_dims == 0 and len(self._nav_shape) == 0:
+                raise ValueError("Shape object is empty")
+            elif self._sig_dims == 0:
+                return Shape(other + self._nav_shape, sig_dims=0)
+            else:
+                return Shape(other + self._sig_shape, sig_dims=len(self._sig_shape + other))
```

Comment:
Hmm. I'm not sure the `sig_dims` behavior should be like this - but let me think a bit more about it. Does this match the usage from your grep survey?

Reply:
Actually, it seems that this is a bit arbitrary. In some cases, `sig_dims` of the final `Shape` is set to the `sig_dims` of the starting `Shape` object but there is at least one instance (src\libertem\io\dataset\base\partition.py line 84) where the number 2 was used. Since `__add__` cannot accept additional parameters, I'm not sure what what's a good solution.

Reply:
> In some cases, sig_dims of the final Shape is set to the sig_dims of the starting Shape object

This may actually be a good default behavior, at least for concatenating something to the left side of a `Shape` (which could be though of as: concatenating to the navigation dimensions). When concatenating a tuple to the right side (which could be thought of as: concatenating to the signal dimensions), changing the `sig_dims` to include the "other" length may be a good idea.

> where the number 2 was used

I think that is actually a bug in `get_macrotile`! The `sig_dims` of the original `Shape` should be used there. /cc @uellue 

> Since __add__ cannot accept additional parameters, I'm not sure what what's a good solution.

Hmm. The solution could also be to not use `+` at all, but instead provide specialized methods for this. If `sig_dims` is completely replaced with a new value, it is okay to keep things as-is, that is, to construct a new `Shape` from scratch:

```python
t_sig = (16, 16)
Shape(tuple(some_shape) + t_sig, sig_dims=2)
```

Or, one could think about an API like this (assuming `shape.sig_dims == 0`, or that we don't care about the previous `sig` part):

```python
t_sig = (16, 16)
s2 = shape.replace_sig(t_sig)  # assuming `shape.sig_dims == 0`
```

But, it is still possible to use `+` here, by throwing away the old `sig` part:

```python
t_sig = (16, 16)
s2 = shape.nav + t_sig  # assuming we set sig_dims to len(t_sig) here
assert s2.sig.dims == 2
```

This pattern tends to come up sometimes:

```python
t_nav = (16, 16)
s2 = Shape(t_nav + tuple(shape.sig), sig_dims=shape.sig.dims)
```

With addition, this would be:

```python
t_nav = (16, 16)
s2 = t_nav + shape.sig  # assuming we keep the original sig_dims of `shape` here
assert s2.sig.dims == shape.sig.dims
```

Or, with a specialized API:

```python
t_nav = (16, 16)
s2 = shape.replace_nav(t_nav)
```

Now, looking at the alternatives, I like the addition variant more than the `replace_*` method. What do you think?

Reply:
@sk1p Addition while keeping the `sig_dims` of the starting `Shape` object also looks best to me. I feel that creating a new `Shape` object is not so hard and is much more flexible than specialized methods like `replace_*`. So I don't think `replace_*` will be used much.

> This may actually be a good default behavior, at least for concatenating something to the left side of a Shape (which could be though of as: concatenating to the navigation dimensions). When concatenating a tuple to the right side (which could be thought of as: concatenating to the signal dimensions), changing the sig_dims to include the "other" length may be a good idea.

This sounds good but it's different from what we started with. In this case, it actually doesn't matter what `sig_dims` and `nav_dims` of the original `Shape` object are. One of them doesn't have to be zero. Whether we add to the signal or navigation dimensions depends only on whether the addition is from the left or from the right. Is that OK?

Reply:
> I feel that creating a new Shape object is not so hard and is much more flexible than specialized methods like replace_*. So I don't think replace_* will be used much.

:+1:

> Whether we add to the signal or navigation dimensions depends only on whether the addition is from the left or from the right. Is that OK?

Yeah, let's try that!

Reply:
@sk1p please take a look now